jwst/associations/lib/rules_level3.py

```diff
@@ -326,3 +326,110 @@ def __init__(self, *args, **kwargs):
 
         # Check and continue initialization.
         super(Asn_NRS_IFU, self).__init__(*args, **kwargs)
+
+
+class Asn_Coron(
+        AsnMixin_OpticalPath,
+        AsnMixin_Base
+):
+    """Coronography
+
+    Notes
+    -----
+
+    Coronography is nearly completely defined by the association candidates
+    produced by APT.
+
+    Tracking Issues:
+
+    - `github #311 <https://github.com/STScI-JWST/jwst/issues/311>`
+    """
+
+    def __init__(self, *args, **kwargs):
+
+        # Setup for checking.
+        self.add_constraints({
+            'exp_type': {
+                'value': (
+                    'nrc_coron'
+                    '|mir_lyot'
+                    '|mir_4qpm'
+                ),
+                'inputs': ['exp_type'],
+                'force_unique': True,
+            },
+            'target': {
+                'value': None,
+                'inputs': ['targetid'],
+                'onlyif': lambda item: self.get_exposure_type(item) == 'science',
+                'force_reprocess': ProcessList.EXISTING,
+            }
+        })
+
+        # PSF is required
+        self.validity.update({
+            'has_psf': {
+                'validated': False,
+                'check': lambda entry: entry['exptype'] == 'psf'
+            }
+        })
+
+        # Check and continue initialization.
+        super(Asn_Coron, self).__init__(*args, **kwargs)
+
+    def _init_hook(self, item):
+        """Post-check and pre-add initialization"""
+
+        self.data['asn_type'] = 'coron3'
+        super(Asn_Coron, self)._init_hook(item)
+
+
+class Asn_AMI(
+        AsnMixin_OpticalPath,
+        AsnMixin_Base
+):
+    """Aperture Mask Interferometry
+
+    Notes
+    -----
+
+    AMI is nearly completely defined by the association candidates
+    produced by APT.
+
+    Tracking Issues:
+
+    - `github #310 <https://github.com/STScI-JWST/jwst/issues/310>`
+    """
+
+    def __init__(self, *args, **kwargs):
+
+        # Setup for checking.
+        self.add_constraints({
+            'exp_type': {
+                'value': 'nis_ami',
+                'inputs': ['exp_type'],
+            },
+            'target': {
+                'value': None,
+                'inputs': ['targetid'],
+                'onlyif': lambda item: self.get_exposure_type(item) == 'science',
+                'force_reprocess': ProcessList.EXISTING,
+            }
+        })
+
+        # PSF is required
```

Comment:
For AMI observations I believe that PSF exposures are not actually required, but are highly advantageous. The calwebb_ami3 pipeline has logic to handle the situation where there aren't any PSF exposures. It can still compute fringe params, etc. for the science target only. It's only coronagraphic that requires PSF exposures.

Reply:
Will handle!